### PR TITLE
graceful SetHealthCheckerHostSet issue #1656

### PR DESCRIPTION
**Pull request by alpha-baby**
### Issues associated with this PR

for issue #1656

### Solutions
You should show your solutions about the issues in your PR, including the overall solutions, details and the changes. At this time, Chinese is allowed to describe these.

### UT result
Unit Test is needed if the code is changed, your unit test should cover boundary cases, corner cases, and some exceptional cases. And you need to show the UT result.

### Benchmark
If your code involves the processing of every request, you should give the Benchmark Result.

### Code Style
+ Make sure `Goimports` has run
+ Show `Golint` result

### CODE REVIEW

**Comment by alpha-baby** [comment]
@nejisama thx,review

File: pkg/upstream/healthcheck/healthchecker.go [hunk 1]
```diff
@@ -124,9 +124,40 @@ func (hc *healthChecker) AddHostCheckCompleteCb(cb types.HealthCheckCb) {
...
+	}
+	for _, deleteHost := range deleteHosts {
+		hc.stopCheck(deleteHost)
+	}
 	hc.hosts = hostSet.Hosts()
-	hc.start()
+}
+
+// findNewAndDeleteHost Find deleted and new host in the updated hostSet
+func findNewAndDeleteHost(old, new []types.Host) ([]types.Host, []types.Host) {
```
Comment (nejisama) [inline pkg/upstream/healthcheck/healthchecker.go:138]: 写一个benchmark 看看这个比较的性能，以前也有类似比较add/delete的逻辑，后来因为性能不太好，改成全量更新了

**Comment by alpha-baby** [comment]
### unit test

```bash
=== RUN   Test_findNewAndDeleteHost
--- PASS: Test_findNewAndDeleteHost (0.00s)
=== RUN   Test_findNewAndDeleteHost/find_delete_and_new
    --- PASS: Test_findNewAndDeleteHost/find_delete_and_new (0.00s)
=== RUN   Test_findNewAndDeleteHost/find_delete
    --- PASS: Test_findNewAndDeleteHost/find_delete (0.00s)
=== RUN   Test_findNewAndDeleteHost/find_new
    --- PASS: Test_findNewAndDeleteHost/find_new (0.00s)
PASS
```

### benchmark

```bash
goos: darwin
goarch: amd64
pkg: mosn.io/mosn/pkg/upstream/healthcheck
Benchmark_findNewAndDeleteHost1/find-1newHost-1deleteHost-from10hosts-12         	 1054143	      1074 ns/op	     614 B/op	       3 allocs/op
Benchmark_findNewAndDeleteHost1/find-10newHost-10deleteHost-from100hosts-12      	  101010	     11905 ns/op	    6400 B/op	      13 allocs/op
Benchmark_findNewAndDeleteHost1/find-10newHost-10deleteHost-from1000hosts-12     	   10000	    114907 ns/op	   82944 B/op	      12 allocs/op
Benchmark_findNewAndDeleteHost1/find-100newHost-100deleteHost-from1000hosts-12   	   10000	    111807 ns/op	   90112 B/op	      18 allocs/op
Benchmark_findNewAndDeleteHost1/find-500newHost-500deleteHost-from1000hosts-12   	   10000	    111299 ns/op	  114688 B/op	      22 allocs/op
PASS
```

File: pkg/upstream/healthcheck/healthchecker.go [hunk 2]
```diff
@@ -124,9 +124,40 @@ func (hc *healthChecker) AddHostCheckCompleteCb(cb types.HealthCheckCb) {
...
+		_, ok := newHostsMap[oldHost.AddressString()]
+		if ok {
+			delete(newHostsMap, oldHost.AddressString())
+		} else {
+			deleteHosts = append(deleteHosts, oldHost)
+		}
+	}
+
+	// find new host
+	newHosts := make([]types.Host, 0)
```
Comment (nejisama) [inline pkg/upstream/healthcheck/healthchecker.go:155]: 这里可以预先分配cap, make([], 0, cap) ，减少append扩容

File: pkg/upstream/healthcheck/healthchecker.go [hunk 3]
```diff
@@ -124,9 +124,40 @@ func (hc *healthChecker) AddHostCheckCompleteCb(cb types.HealthCheckCb) {
 // only called in cluster, lock in cluster
 // SetHealthCheckerHostSet reset the healthchecker's hosts
 func (hc *healthChecker) SetHealthCheckerHostSet(hostSet types.HostSet) {
-	hc.stop()
+	deleteHosts, newHosts := findNewAndDeleteHost(hc.hosts, hostSet.Hosts())
+	for _, newHost := range newHosts {
+		hc.startCheck(newHost)
```
[the remark this answers is not in the document]
Comment (alpha-baby) [reply pkg/upstream/healthcheck/healthchecker.go:129]: 我加了行代码，应该这样就是可以的了把

**Review by nejisama — APPROVED** [review]
LGTM

**Review by taoyuanyuan — APPROVED** [review]
LGTM

**Pull request merged**